pallets/gear/src/internal.rs

```diff
@@ -586,15 +593,45 @@ where
             // Message is going to be inserted into mailbox.
             //
             // No hold bound checks required, because gas_limit isn't less than threshold.
-            if gas_limit >= threshold {
+            to_mailbox = gas_limit >= threshold;
+            if to_mailbox {
                 // Cutting gas for storing in mailbox.
-                GasHandlerOf::<T>::cut(sender_node, dispatch.id(), gas_limit)
+                let gas_amount = gas_for_delay.saturating_add(gas_limit);
+                GasHandlerOf::<T>::cut(sender_node, dispatch.id(), gas_amount)
                     .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
 
                 // TODO: adapt this line if gasful sending appears for reservations (#1828)
                 if let Some(reservation_id) = reservation {
                     Self::remove_gas_reservation_with_task(dispatch.source(), reservation_id);
                 }
+            } else {
+                GasHandlerOf::<T>::cut(sender_node, dispatch.id(), gas_for_delay)
+                    .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+            }
+
+            // `HoldBound` cost builder.
+            let hold_builder = HoldBound::<T>::by(CostsPerBlockOf::<T>::dispatch_stash());
+
+            // Calculating correct hold bound to lock gas
+            let maximal_hold = hold_builder.clone().maximum_for_message(dispatch.id());
+            let bn_delay = delay.saturated_into::<BlockNumberFor<T>>();
+            let hold = hold_builder.duration(bn_delay).min(maximal_hold);
+
+            // Locking funds for holding.
+            GasHandlerOf::<T>::lock(dispatch.id(), hold.lock())
+                .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+
+            if hold.expected_duration().is_zero() {
+                // TODO: Replace with unreachable call after:
+                // - `HoldBound` safety usage stabilized;
+                log::error!("Failed to figure out correct wait hold bound");
```

Comment:
similar stuff `log::error!("Failed to figure out correct wait hold bound");` @ `fn wait_dispatch` is partially unstable now, while this one is unreachable and guaranteed by code above

Reply:
done